File: src/112. Bouncy numbers/112.java
import java.util.Arrays;

public class Main {

    public static void main(String[] args) {
        long bouncyCount = 0;
        long n = 1;
        while (true) {
            if (isBouncy(n)) {
                ++bouncyCount;
            }
            if ((double) bouncyCount / n == 0.99) {
                System.out.println(n);
                break;
            }
            n++;
        }
    }

    private static boolean isBouncy(long n) {
        String original = Long.toString(n);
        String sorted = sortString(original);
        if (original.equals(sorted)) {
            return false;
        }
        sorted = reverseString(sorted);
        if (original.equals(sorted)) {
            return false;
        }
        return true;
    }

    private static String sortString(String original) {
        char[] chars = original.toCharArray();
        Arrays.sort(chars);
        return new String(chars);
    }

    private static String reverseString(String original) {
        return new StringBuilder(original).reverse().toString();
    }

}

File: src/112. Bouncy numbers/112.cpp
#include <iostream>
#include <string>
#include <algorithm>

using std::string;

bool is_bouncy(long long n);

int main() {
    long long bouncy_count = 0;
    long long n = 1;
    while (true) {
        if (is_bouncy(n)) {
            ++bouncy_count;
        }
        if (static_cast<double>(bouncy_count) / n == 0.99) {
            std::cout << n << std::endl;
            break;
        }
        ++n;
    }
}

bool is_bouncy(long long n) {
    string original = std::to_string(n);
    string sorted(original);
    std::sort(sorted.begin(), sorted.end());
    if (original == sorted) {
        return false;
    }
    std::reverse(sorted.begin(), sorted.end());
    if (original == sorted) {
        return false;
    }
    return true;
}
